enhancements/two-nodes-openshift/2no.md

```diff
@@ -0,0 +1,618 @@
+
+
+## Proposal
+
+
+To achieve a two-node OpenShift cluster, we are leveraging traditional high-availability concepts and technologies. The proposed solution includes:
+
+1. Leverage of the Full RHEL-HA Stack:
+   * Run the RHEL-HA stack “under to kubelet” (directly on the hardware, not as an OpenShift workload)
+   * Corosync for super fast failure detection, membership calculations, which in turn will trigger Pacemaker to apply Fencing based on Corosync quorum/membership information.
+   * Pacemaker for integrating membership and quorum information, driving fencing, and managing if/when kubelet and etcd can be started
+   * Pacemaker models kubelet and cri-o as a clone (much like a ReplicaSet) and etcd as a “promotable clone” (think a construct designed for leader/follower style services).  Together with fencing and quorum, this ensures that an isolated node that reboots is inert and can do no harm.
+   * Pacemaker is [configured](//TODO mshitrit add link) to manage etcd/cri-o/kubelet, it will start/stop/restart those services using a script or an executable.
+   * Pacemaker does not understand what it is managing, and expects an executable or script that knows how to start/stop/monitor (and optionally promote/demote) the service.
+     1. Likely we would need to create one for etcd
+     2. For kubelet and cri-o we can likely use the existing systemd unit file
+2. Failure Scenarios:
+   * Implement detailed handling procedures for cold boots, network failures, node failures, kubelet failures, and etcd failures using the RHEL-HA stack.
+    [see examples](#failure-handling)
+3. Fencing Methods:
+   * We plan to use Baseboard Management Controller (BMC) as our primary fencing method, the premise of using BMC for fencing is that a node that is powered off, or was previously powered off and configured to be inert until quorum forms, is not in a position to cause corruption or diverging datasets.  Sending power-off (or reboot) commands to the peer’s BMC achieves this goal.
+
+
+### Workflow Description
+
+#### Cluster Creator Role:
+* The Cluster Creator will automatically install the 2NO (by using an installer), installation process will include the following steps:  
+  * Deploys a two-node OpenShift cluster
+  * Configures cluster membership and quorum using Corosync.
+  * Sets up Pacemaker for resource management and fencing.
+
+#### Application Administrator Role:
+* Receives cluster credentials.
+* Deploys applications within the two-node cluster environment.
+
+#### Failure Handling:
+
+1. Cold Boot
+   1. One node (Node1) boots
+   2. Node1 does have “corosync quorum” (C-quorum)  (requires forming a membership with it’s peer)
+   3. Node1 does not start etcd or kubelet, remains inert waiting for Node2
+   4. Peer (Node2) boots
+   5. Corosync membership containing both nodes forms
+   6. Pacemaker “starts” etcd on both nodes
```

Comment:
To make it clearer, we could draw a line between the actual `etcd` and the Pacemaker's `resource agent` (RA) which handles the `etcd` instance's life cycle.

Pacemaker can `start` the RA (calling its `start` function), without actually starting the `etcd` instance, but inspecting the current status instead (e.g. the revision number), in order to decide how to start the `etcd` process (as leader, as learner, etc.)

Reply:
At this point the RA's promotion can correspond to the actual start of `etcd` instance

Reply:
Good point, but IIUC the gist of what you are saying in this particular context the distinction between RA/Pacemaker isn't necessary.

Reply:
Indeed, in this context we can just say "Pacemaker starts etcd".
Quoting the word _starts_ creates ambiguity about its intended meaning.